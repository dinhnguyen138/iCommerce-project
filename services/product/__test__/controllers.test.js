'use strict';
const controllers = require('../controllers');
const services = require('../services');
jest.mock('../services');

const _data = [
  {
    _id: '0',
    name: 'Macbook air 2018',
    categories: ['Laptop', 'Electronic', 'Apple'],
    description: 'Macbook new design 2018',
    rating: 5,
    originalprice: 50000,
    currentprice: 30000,
    brand: 'Apple',
    seller: 'FPT'
  },
  {
    _id: '1',
    name: 'HP Notebook 2020',
    categories: ['Laptop', 'Electronic', 'Device'],
    description: 'Brand new device notebook from HP',
    rating: 4.5,
    originalprice: 45000,
    currentprice: 35000,
    brand: 'HP',
    seller: 'VienThongA'
  },
  {
    _id: '2',
    name: 'Cannon D500',
    categories: ['Camera', 'Electronic', 'Device'],
    description: 'New camera 2020',
    rating: 4.5,
    originalprice: 40000,
    currentprice: 38000,
    brand: 'Cannon',
    seller: 'FPT'
  }
];
services.getProducts.mockImplementation(async query => {
  console.log(query);
  if ( query.query === 'hp') {
    return [_data[1]];
  }
  if ( query.query === 'Cannon' ) {
    return [_data[2]];
  }
  if (query.brand === 'Cannon' ) {
    return [_data[2]];
  }
  if (query.category === 'Electronic') {
    return [_data[0], _data[1], _data[2]];
  }
  if (query.rating === 4) {
    return [_data[0], _data[1], _data[2]];
  }
  if (query.price === '20000:30000') {
    return [_data[0]];
  }
  if (query.sort === 'seller:asc' ) {
    return [_data[0], _data[2], _data[1]];
  }
  return [];
});
services.getProduct.mockImplementation(id => {
  return _data[parseInt(id)];
})
services.createProduct.mockImplementation(value => {
  if (value.simulate === 'ok') {
    value.id = '1';
    delete value.simulate;
  }
  if (value.simulate === 'failed') {
    value = null;
  }
  return value;
});
describe('Controllers test', () => {
  test('getProducts test', async () => {
    let ctx = {
      request: {},
      params: {},
      body: '',
      throw: jest.fn(() => {})
    };
    let generateRequest = content => {
      ctx.request.query = content;
      return ctx;
    };
    await controllers.getProducts(generateRequest({ query: 'hp' }));
    expect(ctx.body).toMatchSnapshot();
    await controllers.getProducts(generateRequest({ brand: 'Cannon'}));
    expect(ctx.body).toMatchSnapshot();
    await controllers.getProducts(generateRequest({ category: 'Electronic' }));
    expect(ctx.body).toMatchSnapshot();
    await controllers.getProducts(generateRequest({ rating: 4 }));
    expect(ctx.body).toMatchSnapshot();
    await controllers.getProducts(generateRequest({ price: '20000:30000' }));
    expect(ctx.body).toMatchSnapshot();
    await controllers.getProducts(generateRequest({ sort: 'seller:asc'}));
    expect(ctx.body).toMatchSnapshot();
  }); 
  test('getProduct test', async () => {
    let ctx = {
      params: {},
      request: {},
      body: '',
      throw: jest.fn(() => {})
    };
    let addParams = id => {
      ctx.params.id = id;
      return ctx;
    };
    console.log(ctx);
    await controllers.getProduct(addParams(1));
    expect(ctx.body).toMatchSnapshot();
    await controllers.getProduct(addParams(5));
    expect(ctx.throw).toHaveBeenCalledTimes(1);
  });
  test('createProduct test', async () => {
    let ctx = {
      params: {},
      request: {},
      body: '',
      throw: jest.fn(() => {})
    };
    let data = {
      name: 'HP Notebook 2020',
      categories: ['Laptop', 'Electronic', 'Device'],
      description: 'Brand new device notebook from HP',
      rating: 4.5,
      originalprice: 45000
    };
    let addBody = (body, simulate) => {
      ctx.request.body = body;
      ctx.request.body.simulate = simulate;
      return ctx;
    };
    await controllers.createProduct(addBody(data, 'ok'));
    expect(ctx.body).toMatchSnapshot();
    await controllers.createProduct(addBody(data, 'failed'));
    expect(ctx.throw).toHaveBeenCalled();
    await controllers.createProduct(addBody(data, 'xxxx'));
    expect(ctx.throw).toHaveBeenCalled();
  });
});