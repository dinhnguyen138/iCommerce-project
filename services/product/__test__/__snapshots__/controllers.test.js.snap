// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controllers test createProduct test 1`] = `
Object {
  "categories": Array [
    "Laptop",
    "Electronic",
    "Device",
  ],
  "description": "Brand new device notebook from HP",
  "id": "1",
  "name": "HP Notebook 2020",
  "originalprice": 45000,
  "rating": 4.5,
}
`;

exports[`Controllers test getProduct test 1`] = `
Object {
  "_id": "1",
  "brand": "HP",
  "categories": Array [
    "Laptop",
    "Electronic",
    "Device",
  ],
  "currentprice": 35000,
  "description": "Brand new device notebook from HP",
  "name": "HP Notebook 2020",
  "originalprice": 45000,
  "rating": 4.5,
  "seller": "VienThongA",
}
`;

exports[`Controllers test getProducts test 1`] = `
Object {
  "brands": undefined,
  "categories": undefined,
  "products": Array [
    Object {
      "_id": "1",
      "brand": "HP",
      "categories": Array [
        "Laptop",
        "Electronic",
        "Device",
      ],
      "currentprice": 35000,
      "description": "Brand new device notebook from HP",
      "name": "HP Notebook 2020",
      "originalprice": 45000,
      "rating": 4.5,
      "seller": "VienThongA",
    },
  ],
  "ratings": undefined,
  "sellers": undefined,
}
`;

exports[`Controllers test getProducts test 2`] = `
Object {
  "brands": undefined,
  "categories": undefined,
  "products": Array [
    Object {
      "_id": "2",
      "brand": "Cannon",
      "categories": Array [
        "Camera",
        "Electronic",
        "Device",
      ],
      "currentprice": 38000,
      "description": "New camera 2020",
      "name": "Cannon D500",
      "originalprice": 40000,
      "rating": 4.5,
      "seller": "FPT",
    },
  ],
  "ratings": undefined,
  "sellers": undefined,
}
`;

exports[`Controllers test getProducts test 3`] = `
Object {
  "brands": undefined,
  "categories": undefined,
  "products": Array [
    Object {
      "_id": "0",
      "brand": "Apple",
      "categories": Array [
        "Laptop",
        "Electronic",
        "Apple",
      ],
      "currentprice": 30000,
      "description": "Macbook new design 2018",
      "name": "Macbook air 2018",
      "originalprice": 50000,
      "rating": 5,
      "seller": "FPT",
    },
    Object {
      "_id": "1",
      "brand": "HP",
      "categories": Array [
        "Laptop",
        "Electronic",
        "Device",
      ],
      "currentprice": 35000,
      "description": "Brand new device notebook from HP",
      "name": "HP Notebook 2020",
      "originalprice": 45000,
      "rating": 4.5,
      "seller": "VienThongA",
    },
    Object {
      "_id": "2",
      "brand": "Cannon",
      "categories": Array [
        "Camera",
        "Electronic",
        "Device",
      ],
      "currentprice": 38000,
      "description": "New camera 2020",
      "name": "Cannon D500",
      "originalprice": 40000,
      "rating": 4.5,
      "seller": "FPT",
    },
  ],
  "ratings": undefined,
  "sellers": undefined,
}
`;

exports[`Controllers test getProducts test 4`] = `
Object {
  "brands": undefined,
  "categories": undefined,
  "products": Array [
    Object {
      "_id": "0",
      "brand": "Apple",
      "categories": Array [
        "Laptop",
        "Electronic",
        "Apple",
      ],
      "currentprice": 30000,
      "description": "Macbook new design 2018",
      "name": "Macbook air 2018",
      "originalprice": 50000,
      "rating": 5,
      "seller": "FPT",
    },
    Object {
      "_id": "1",
      "brand": "HP",
      "categories": Array [
        "Laptop",
        "Electronic",
        "Device",
      ],
      "currentprice": 35000,
      "description": "Brand new device notebook from HP",
      "name": "HP Notebook 2020",
      "originalprice": 45000,
      "rating": 4.5,
      "seller": "VienThongA",
    },
    Object {
      "_id": "2",
      "brand": "Cannon",
      "categories": Array [
        "Camera",
        "Electronic",
        "Device",
      ],
      "currentprice": 38000,
      "description": "New camera 2020",
      "name": "Cannon D500",
      "originalprice": 40000,
      "rating": 4.5,
      "seller": "FPT",
    },
  ],
  "ratings": undefined,
  "sellers": undefined,
}
`;

exports[`Controllers test getProducts test 5`] = `
Object {
  "brands": undefined,
  "categories": undefined,
  "products": Array [
    Object {
      "_id": "0",
      "brand": "Apple",
      "categories": Array [
        "Laptop",
        "Electronic",
        "Apple",
      ],
      "currentprice": 30000,
      "description": "Macbook new design 2018",
      "name": "Macbook air 2018",
      "originalprice": 50000,
      "rating": 5,
      "seller": "FPT",
    },
  ],
  "ratings": undefined,
  "sellers": undefined,
}
`;

exports[`Controllers test getProducts test 6`] = `
Object {
  "brands": undefined,
  "categories": undefined,
  "products": Array [
    Object {
      "_id": "0",
      "brand": "Apple",
      "categories": Array [
        "Laptop",
        "Electronic",
        "Apple",
      ],
      "currentprice": 30000,
      "description": "Macbook new design 2018",
      "name": "Macbook air 2018",
      "originalprice": 50000,
      "rating": 5,
      "seller": "FPT",
    },
    Object {
      "_id": "2",
      "brand": "Cannon",
      "categories": Array [
        "Camera",
        "Electronic",
        "Device",
      ],
      "currentprice": 38000,
      "description": "New camera 2020",
      "name": "Cannon D500",
      "originalprice": 40000,
      "rating": 4.5,
      "seller": "FPT",
    },
    Object {
      "_id": "1",
      "brand": "HP",
      "categories": Array [
        "Laptop",
        "Electronic",
        "Device",
      ],
      "currentprice": 35000,
      "description": "Brand new device notebook from HP",
      "name": "HP Notebook 2020",
      "originalprice": 45000,
      "rating": 4.5,
      "seller": "VienThongA",
    },
  ],
  "ratings": undefined,
  "sellers": undefined,
}
`;
